age-plugin.md

```diff
@@ -0,0 +1,541 @@
+
+Users interact with age clients by providing either recipients for file
+encryption, or identities for file decryption. When a plugin recipient or
+identity is provided, the age client searches the `PATH` for a binary with the
+corresponding plugin name.
+
+Recipient stanza types are not required to be correlated to specific plugin
+names. When decrypting, age clients will pass all recipient stanzas to every
+connected plugin. Plugins MUST ignore stanzas that they do not know about.
+
+A plugin binary may handle multiple recipient or identity types by being present
+in the `PATH` under multiple names. This can be implemented with symlinks or
+aliases to the canonical binary.
+
+Multiple plugin binaries can support the same recipient and identity types; the
+first binary found in the `PATH` will be used by age clients. Some Unix OSs
+support "alternatives", which plugin binaries should leverage if they provide
+support for a common recipient or identity type.
+
+Note that the identity specified by a user doesn't need to point to a specific
+decryption key, or indeed contain any key material at all. It only needs to
+contain sufficient information for the plugin to locate the necessary key
+material.
+
+### Standard age keys
+
+A plugin MAY support decrypting files encrypted to native age recipients, by
+including support for the `x25519` recipient stanza. Such plugins will pick
+their own name, and users will use identity files containing identities that
+specify that plugin name.
+
+### Agents for age identities
+
+One use-case for plugins is for implementing agents for age identities. The
+expected design is a short-lifetime plugin binary `age-plugin-NAME`, which
+implements the plugin protocol, and in turn connects to (or starts, if not
+already running) a long-lifetime agent daemon. Specification of agents is
+out of scope for this document.
+
+## State machines
+
+A plugin operates using one of several state machines. The age client chooses
+which state machine to use when it starts the plugin, by passing an argument
+flag `--age-plugin=STATE_MACHINE` which specifies the type and version of the
+state machine. This document defines the following state machines:
+
+- `recipient-v1` - for wrapping file keys during file encryption.
+- `identity-v1` - for unwrapping file keys during file decryption.
+
+Plugins MUST NOT make any assumptions about the working directory that they are
+run inside.
+
+A plugin MUST refuse to start if it does not know about the requested state
+machine; in this situation, or if a plugin otherwise terminates early with an
+error, the age client:
+
+- MUST propagate the failure to the user if it occurs during file encryption
+  (which would mean that the file could not be encrypted to one of the requested
+  recipients). This may include displaying the contents of the plugin's standard
+  error to the user.
+- MAY ignore the failure if it occurs during file decryption (and try another
+  identity or plugin).
```

Comment:
It's unclear to me what "refuse to start" looks like here; does it involve replying with a protocol-structured error, or just unceremoniously dumping whatever to stderr and exiting with an error status code? The text implies the latter, but rage at least reacts to that with a rather unhelpful "Error: incomplete response", though I realize this could be an implementation infelicity.

Reply:
The latter is intended here; if the requested state machine is unknown then there is no guarantee that the IPC protocol it uses is compatible with any plugin IPC that the client speaks. Note that the stanza-based IPC protocol is specified below this point, and is documented as for "v1 state machines".

> rage at least reacts to that with a rather unhelpful "Error: incomplete response", though I realize this could be an implementation infelicity.

Indeed, this is something I should fix in `rage` 😅

Reply:
That makes sense, thanks! Part of my uncertainty was that `age_plugin::run_state_machine` doesn't offer any way to handle only `identity-v1` (you can't `panic!` in the other branch because you still need a concrete `R: RecipientPluginV1` to type it with) other than implementing the whole interface and returning protocol errors inside; I suppose there is a distinction to be made between *unknown* state machines and *known but inapplicable* ones?

Reply:
That's indeed the case. It's usually the case that you'll want to implement both (for entirely custom recipients), but indeed for a plugin that handles native age identities you wouldn't need the recipient state machine. Maybe there are cases where you might only want to implement the recipient state machine, but I can't think of any at present.

In any case, this should be fixed in the `age-plugin` crate (probably with the addition of a `run_identity_state_machine` function).

Reply:
Opened https://github.com/str4d/rage/issues/387.